thunder/dynamo/compiler.py

```diff
@@ -127,3 +130,32 @@ def save_reproducer_to_folder(self, reproducer_folder: str | PathLike, use_pytes
                     f"graph{graph_idx}_{cur_name}",
                     use_pytest_benchmark,
                 )
+
+
+def thunderfx(fn: Callable, /, **kwargs) -> Callable:
+    """Compiles a callable (function or model) by using Thunder as the backend of :func:`torch.compile`
+    Args:
+        fn: A :class:`~torch.nn.Module` or a function to compile.
+    Keyword Args:
+        **kwargs: a dictionary of options to pass to :func:`torch.compile` and :func:`thunder.jit`.
+    Returns:
+        The compiled callable
+    """
+    import thunder
+
+    torch_compile_kwarg_names = inspect.getfullargspec(torch.compile).kwonlyargs
+    thunder_jit_kwarg_names = inspect.getfullargspec(thunder.jit).kwonlyargs
+    overlap = [kwarg_name for kwarg_name in thunder_jit_kwarg_names if kwarg_name in torch_compile_kwarg_names]
+    check(
+        not overlap,
+        lambda: f"There are overlapping kwargs between thunder.jit and torch.compile: {overlap}",
+        ValueError,
+    )
+
+    torch_compile_options = {k: v for k, v in kwargs.items() if k in torch_compile_kwarg_names}
+    thunder_options = {k: v for k, v in kwargs.items() if k not in torch_compile_kwarg_names}
+
+    backend = ThunderCompiler(**thunder_options)
```

Comment:
When the `ThunderCompiler` class is created this way, it sets `self._torch_compile` here:

https://github.com/Lightning-AI/lightning-thunder/blob/6307021dbc7fa73f589ac72e2f3deea4c5cabb25/thunder/dynamo/compiler.py#L77

The `self._torch_compile` object is used here:

https://github.com/Lightning-AI/lightning-thunder/blob/6307021dbc7fa73f589ac72e2f3deea4c5cabb25/thunder/dynamo/compiler.py#L88

But with this approach, won't the torch.compile options not be passed to ThunderCompiler, so its internal `self._torch_compile` object won't be set properly?

Does the call to create the `ThunderCompiler` object need to look like `ThunderCompiler(thunder_options, torch_compile_options)`, and then internally it needs to call

```
self._torch_compile = partial(torch.compile, **torch_compile_options)
```

?

Reply:
(Maybe in the future we don't want the "outer" torch.compile kwargs to be the same as the "inner" torch.compile kwargs, so we might want to leave a note about that)

Reply:
> But with this approach, won't the torch.compile options not be passed to ThunderCompiler, so its internal self._torch_compile object won't be set properly?

I think currently the 2 groups of torch.compile kwargs can be different, one for the outside `torch.compile`, it's mainly used for the dynamo part I think, the other group is used in the inner `torch.compile` to handle the subgraph fall back to inductor

```
backend = ThunderCompiler(torch_inductor_options={dynamic=True})
cfunc = torch.compile(func, backend=backend, dynamic=False)
```
cc: @kshitij12345 do I understand it right?

Reply:
I understand, and I think you're right. But what does that look like with the new `thunderfx` function?

```
thunderfx(foo, full_graph=True, torch_inductor_options={'full_graph'=True})
```

?

That seems a little odd and hidden.

I would think we should do one of two things for now:

1. Set the torch.compile kwargs consistently (both "outer" and "inner" are the same)
2. Set the "inner" torch.compile kwargs however we like, without providing a mechanism for practitioners to set them (at this time)

What do you think, @kshitij12345?

Reply:
Right, now we do support `torch.compile` inside the `ThunderCompiler` have different arguments than the user one (outer). However, I don't think in practice, we had to set them differently yet.

So, I like the option 2 which seems less confusing.

NOTE - Even though we call `torch.compile` inside in `ThunderCompiler`, technically we just want something like `torch._inductor.compile_fx` (as the dynamo stage is already done).

Reply:
Then in that case the current code will work, the `torch.compile` kwargs passed to `thunder_fx` will only set the outer `torch.compile` and not affect the inner ones.

Reply:
@kshitij12345 Do we want to remove this `torch_inductor_options` option completely in the constructor of `ThunderCompiler`?

Reply:
Sure, we can remove it.